{
  "cells": [
    {
      "cell_type": "code",
      "execution_count": null,
      "metadata": {
        "id": "3TNZMm2Mbn6l"
      },
      "outputs": [],
      "source": [
        "!git clone https://huggingface.co/spaces/PascalNotin/Tranception_design\n",
        "!mv Tranception_design/* ./\n",
        "!rm -r Tranception_design\n",
        "!pip install -r requirements.txt"
      ]
    },
    {
      "cell_type": "code",
      "execution_count": null,
      "metadata": {
        "id": "m4xeDX3RcCY8"
      },
      "outputs": [],
      "source": [
        "import torch\n",
        "import transformers\n",
        "from transformers import PreTrainedTokenizerFast\n",
        "import tranception\n",
        "import datasets\n",
        "from tranception import config, model_pytorch\n",
        "import numpy as np\n",
        "import pandas as pd\n",
        "import matplotlib.pyplot as plt\n",
        "import seaborn as sns\n",
        "import gradio as gr\n",
        "\n",
        "AA_vocab = \"ACDEFGHIKLMNPQRSTVWY\"\n",
        "tokenizer = PreTrainedTokenizerFast(tokenizer_file=\"./tranception/utils/tokenizers/Basic_tokenizer\",\n",
        "                                                unk_token=\"[UNK]\",\n",
        "                                                sep_token=\"[SEP]\",\n",
        "                                                pad_token=\"[PAD]\",\n",
        "                                                cls_token=\"[CLS]\",\n",
        "                                                mask_token=\"[MASK]\"\n",
        "                                            )\n",
        "\n",
        "def create_all_single_mutants(sequence,AA_vocab=AA_vocab,mutation_range_start=None,mutation_range_end=None):\n",
        "  all_single_mutants={}\n",
        "  sequence_list=list(sequence)\n",
        "  if mutation_range_start is None: mutation_range_start=1\n",
        "  if mutation_range_end is None: mutation_range_end=len(sequence)\n",
        "  for position,current_AA in enumerate(sequence[mutation_range_start-1:mutation_range_end]):\n",
        "    for mutated_AA in AA_vocab:\n",
        "      if current_AA!=mutated_AA:\n",
        "        mutated_sequence = sequence_list.copy()\n",
        "        mutated_sequence[position] = mutated_AA\n",
        "        all_single_mutants[current_AA+str(mutation_range_start+position)+mutated_AA]=\"\".join(mutated_sequence)\n",
        "  all_single_mutants = pd.DataFrame.from_dict(all_single_mutants,columns=['mutated_sequence'],orient='index')\n",
        "  all_single_mutants.reset_index(inplace=True)\n",
        "  all_single_mutants.columns = ['mutant','mutated_sequence']\n",
        "  return all_single_mutants\n",
        "\n",
        "def create_scoring_matrix_visual(scores,sequence,image_index=0,mutation_range_start=None,mutation_range_end=None,AA_vocab=AA_vocab,annotate=True,fontsize=20):\n",
        "  filtered_scores=scores.copy()\n",
        "  filtered_scores=filtered_scores[filtered_scores.position.isin(range(mutation_range_start,mutation_range_end+1))]\n",
        "  piv=filtered_scores.pivot(index='position',columns='target_AA',values='avg_score').round(4)\n",
        "  mutation_range_len = mutation_range_end - mutation_range_start + 1\n",
        "  fig, ax = plt.subplots(figsize=(50,mutation_range_len))\n",
        "  scores_dict = {}\n",
        "  valid_mutant_set=set(filtered_scores.mutant)  \n",
        "  ax.tick_params(bottom=True, top=True, left=True, right=True)\n",
        "  ax.tick_params(labelbottom=True, labeltop=True, labelleft=True, labelright=True)\n",
        "  if annotate:\n",
        "    for position in range(mutation_range_start,mutation_range_end+1):\n",
        "      for target_AA in list(AA_vocab):\n",
        "        mutant = sequence[position-1]+str(position)+target_AA\n",
        "        if mutant in valid_mutant_set:\n",
        "          scores_dict[mutant]= float(filtered_scores.loc[filtered_scores.mutant==mutant,'avg_score'])\n",
        "        else:\n",
        "          scores_dict[mutant]=0.0\n",
        "    labels = (np.asarray([\"{} \\n {:.4f}\".format(symb,value) for symb, value in scores_dict.items() ])).reshape(mutation_range_len,len(AA_vocab))\n",
        "    heat = sns.heatmap(piv,annot=labels,fmt=\"\",cmap='RdYlGn',linewidths=0.30,ax=ax,vmin=np.percentile(scores.avg_score,2),vmax=np.percentile(scores.avg_score,98),\\\n",
        "                cbar_kws={'label': 'Log likelihood ratio (mutant / starting sequence)'},annot_kws={\"size\": fontsize})\n",
        "  else:\n",
        "    heat = sns.heatmap(piv,cmap='RdYlGn',linewidths=0.30,ax=ax,vmin=np.percentile(scores.avg_score,2),vmax=np.percentile(scores.avg_score,98),\\\n",
        "                cbar_kws={'label': 'Log likelihood ratio (mutant / starting sequence)'},annot_kws={\"size\": fontsize})\n",
        "  heat.figure.axes[-1].yaxis.label.set_size(fontsize=int(fontsize*1.5))\n",
        "  heat.figure.axes[-1].yaxis.set_ticklabels(heat.figure.axes[-1].yaxis.get_ticklabels(), fontsize=fontsize)\n",
        "  heat.set_title(\"Higher predicted scores (green) imply higher protein fitness\",fontsize=fontsize*2, pad=40)\n",
        "  heat.set_ylabel(\"Sequence position\", fontsize = fontsize*2)\n",
        "  heat.set_xlabel(\"Amino Acid mutation\", fontsize = fontsize*2)\n",
        "  yticklabels = [str(pos)+' ('+sequence[pos-1]+')' for pos in range(mutation_range_start,mutation_range_end+1)]\n",
        "  heat.set_yticklabels(yticklabels)\n",
        "  heat.set_xticklabels(heat.get_xmajorticklabels(), fontsize = fontsize)\n",
        "  heat.set_yticklabels(heat.get_ymajorticklabels(), fontsize = fontsize, rotation=0)\n",
        "  plt.tight_layout()\n",
        "  image_path = 'fitness_scoring_substitution_matrix_{}.png'.format(image_index)\n",
        "  plt.savefig(image_path,dpi=100)\n",
        "  plt.show()\n",
        "  return image_path\n",
        "\n",
        "def suggest_mutations(scores):\n",
        "  intro_message = \"The following mutations may be sensible options to improve fitness: \\n\\n\"\n",
        "  #Best mutants\n",
        "  top_mutants=list(scores.sort_values(by=['avg_score'],ascending=False).head(5).mutant)\n",
        "  top_mutants_fitness=list(scores.sort_values(by=['avg_score'],ascending=False).head(5).avg_score)\n",
        "  top_mutants_recos = [top_mutant+\" (\"+str(round(top_mutant_fitness,4))+\")\" for (top_mutant,top_mutant_fitness) in zip(top_mutants,top_mutants_fitness)]\n",
        "  mutant_recos = \"The single mutants with highest predicted fitness are (positive scores indicate fitness increase Vs starting sequence, negative scores indicate fitness decrease):\\n {} \\n\\n\".format(\", \".join(top_mutants_recos))\n",
        "  #Best positions\n",
        "  positive_scores = scores[scores.avg_score > 0]\n",
        "  positive_scores_position_avg = positive_scores.groupby(['position']).mean()\n",
        "  top_positions=list(positive_scores_position_avg.sort_values(by=['avg_score'],ascending=False).head(5).index.astype(str))\n",
        "  position_recos = \"The positions with the highest average fitness increase are (only positions with at least one fitness increase are considered):\\n {}\".format(\", \".join(top_positions))\n",
        "  return intro_message+mutant_recos+position_recos\n",
        "\n",
        "def check_valid_mutant(sequence,mutant,AA_vocab=AA_vocab):\n",
        "  valid = True\n",
        "  try:\n",
        "    from_AA, position, to_AA = mutant[0], int(mutant[1:-1]), mutant[-1]\n",
        "  except:\n",
        "    valid = False\n",
        "  if sequence[position-1]!=from_AA: valid=False\n",
        "  if position<1 or position>len(sequence): valid=False\n",
        "  if to_AA not in AA_vocab: valid=False\n",
        "  return valid\n",
        "\n",
        "def get_mutated_protein(sequence,mutant):\n",
        "  assert check_valid_mutant(sequence,mutant), \"The mutant is not valid\"\n",
        "  mutated_sequence = list(sequence)\n",
        "  mutated_sequence[int(mutant[1:-1])-1]=mutant[-1]\n",
        "  return ''.join(mutated_sequence)\n",
        "\n",
        "def score_and_create_matrix_all_singles(sequence,mutation_range_start=None,mutation_range_end=None,model_type=\"Small\",scoring_mirror=False,batch_size_inference=20,max_number_positions_per_heatmap=50,num_workers=0,AA_vocab=AA_vocab):\n",
        "  if mutation_range_start is None: mutation_range_start=1\n",
        "  if mutation_range_end is None: mutation_range_end=len(sequence)\n",
        "  assert len(sequence) > 0, \"no sequence entered\"\n",
        "  assert mutation_range_start <= mutation_range_end, \"mutation range is invalid\"\n",
        "  if model_type==\"Small\":\n",
        "    model = tranception.model_pytorch.TranceptionLMHeadModel.from_pretrained(pretrained_model_name_or_path=\"PascalNotin/Tranception_Small\")\n",
        "  elif model_type==\"Medium\":\n",
        "    model = tranception.model_pytorch.TranceptionLMHeadModel.from_pretrained(pretrained_model_name_or_path=\"PascalNotin/Tranception_Medium\")\n",
        "  elif model_type==\"Large\":\n",
        "    model = tranception.model_pytorch.TranceptionLMHeadModel.from_pretrained(pretrained_model_name_or_path=\"PascalNotin/Tranception_Large\")\n",
        "  if torch.cuda.is_available():\n",
        "    model.cuda()\n",
        "    print(\"Inference will take place on GPU\")\n",
        "  else:\n",
        "    print(\"Inference will take place on CPU\")\n",
        "  model.config.tokenizer = tokenizer\n",
        "  all_single_mutants = create_all_single_mutants(sequence,AA_vocab,mutation_range_start,mutation_range_end)\n",
        "  scores = model.score_mutants(DMS_data=all_single_mutants, \n",
        "                                    target_seq=sequence, \n",
        "                                    scoring_mirror=scoring_mirror, \n",
        "                                    batch_size_inference=batch_size_inference,  \n",
        "                                    num_workers=num_workers, \n",
        "                                    indel_mode=False\n",
        "                                    )\n",
        "  scores = pd.merge(scores,all_single_mutants,on=\"mutated_sequence\",how=\"left\")\n",
        "  scores[\"position\"]=scores[\"mutant\"].map(lambda x: int(x[1:-1]))\n",
        "  scores[\"target_AA\"] = scores[\"mutant\"].map(lambda x: x[-1])\n",
        "  score_heatmaps = []\n",
        "  mutation_range = mutation_range_end - mutation_range_start + 1\n",
        "  number_heatmaps = int((mutation_range - 1) / max_number_positions_per_heatmap) + 1\n",
        "  image_index = 0\n",
        "  window_start = mutation_range_start\n",
        "  window_end = min(mutation_range_end,mutation_range_start+max_number_positions_per_heatmap-1)\n",
        "  for image_index in range(number_heatmaps):\n",
        "    score_heatmaps.append(create_scoring_matrix_visual(scores,sequence,image_index,window_start,window_end,AA_vocab))\n",
        "    window_start += max_number_positions_per_heatmap\n",
        "    window_end = min(mutation_range_end,window_start+max_number_positions_per_heatmap-1)\n",
        "  return score_heatmaps, suggest_mutations(scores)\n",
        "\n",
        "def extract_sequence(example):\n",
        "  label, taxon, sequence = example\n",
        "  return sequence\n",
        "\n",
        "def clear_inputs(protein_sequence_input,mutation_range_start,mutation_range_end):\n",
        "  protein_sequence_input = \"\"\n",
        "  mutation_range_start = None\n",
        "  mutation_range_end = None\n",
        "  return protein_sequence_input,mutation_range_start,mutation_range_end"
      ]
    },
    {
      "cell_type": "code",
      "execution_count": null,
      "metadata": {
        "id": "XxoLMOwD0E_b"
      },
      "outputs": [],
      "source": [
        "tranception_design = gr.Blocks()\n",
        "\n",
        "with tranception_design:\n",
        "    gr.Markdown(\"# In silico directed evolution for protein redesign with Tranception\")\n",
        "    gr.Markdown(\" Perform in silico directed evolution with Tranception to iteratively improve the fitness of a protein of interest, one mutation at a time. At each step, the Tranception model computes the log likelihood ratios of all possible single amino acid substitution Vs the starting sequence, and outputs a fitness heatmap and recommandations to guide the selection of the mutation to apply.\")\n",
        "    \n",
        "    with gr.Tabs():\n",
        "        with gr.TabItem(\"Input\"):\n",
        "            with gr.Row():\n",
        "                protein_sequence_input = gr.Textbox(lines=1, \n",
        "                                                label=\"Protein sequence\",\n",
        "                                                placeholder = \"Input the sequence of amino acids representing the starting protein of interest or select one from the list of examples below. You may enter the full sequence or just a subdomain (providing full context typically leads to better results, but is slower at inference)\"\n",
        "                                                )\n",
        "            \n",
        "            with gr.Row():\n",
        "                mutation_range_start = gr.Number(label=\"Start of mutation window (first position indexed at 1)\", value=1, precision=0)\n",
        "                mutation_range_end = gr.Number(label=\"End of mutation window (leave empty for full lenth)\", value=10, precision=0)\n",
        "\n",
        "        with gr.TabItem(\"Parameters\"):\n",
        "            with gr.Row():\n",
        "                model_size_selection = gr.Radio(label=\"Tranception model size (larger models are more accurate but are slower at inference)\", \n",
        "                                                choices=[\"Small\",\"Medium\",\"Large\"], \n",
        "                                                value=\"Small\")\n",
        "            with gr.Row():\n",
        "                scoring_mirror = gr.Checkbox(label=\"Score protein from both directions (leads to more robust fitness predictions, but doubles inference time)\")\n",
        "            with gr.Row():\n",
        "                batch_size_inference = gr.Number(label=\"Model batch size at inference time\",value = 100, precision=0)\n",
        "            with gr.Row():\n",
        "                gr.Markdown(\"Note: the current version does not leverage retrieval of homologs at inference time to increase fitness prediction performance.\")\n",
        "        with gr.Row():\n",
        "            clear_button = gr.Button(value=\"Clear\",variant=\"secondary\")\n",
        "            run_button = gr.Button(value=\"Predict fitness\",variant=\"primary\")\n",
        "    protein_ID = gr.Textbox(label=\"Uniprot ID\", visible=False)\n",
        "    taxon = gr.Textbox(label=\"Taxon\", visible=False)\n",
        "    examples = gr.Examples(\n",
        "        inputs=[protein_ID, taxon, protein_sequence_input],\n",
        "        outputs=[protein_sequence_input],\n",
        "        fn=extract_sequence,\n",
        "        examples=[\n",
        "            ['ADRB2_HUMAN'  ,'Human',           'MGQPGNGSAFLLAPNGSHAPDHDVTQERDEVWVVGMGIVMSLIVLAIVFGNVLVITAIAKFERLQTVTNYFITSLACADLVMGLAVVPFGAAHILMKMWTFGNFWCEFWTSIDVLCVTASIETLCVIAVDRYFAITSPFKYQSLLTKNKARVIILMVWIVSGLTSFLPIQMHWYRATHQEAINCYANETCCDFFTNQAYAIASSIVSFYVPLVIMVFVYSRVFQEAKRQLQKIDKSEGRFHVQNLSQVEQDGRTGHGLRRSSKFCLKEHKALKTLGIIMGTFTLCWLPFFIVNIVHVIQDNLIRKEVYILLNWIGYVNSGFNPLIYCRSPDFRIAFQELLCLRRSSLKAYGNGYSSNGNTGEQSGYHVEQEKENKLLCEDLPGTEDFVGHQGTVPSDNIDSQGRNCSTNDSLL'],\n",
        "            ['IF1_ECOLI'    ,'Prokaryote',      'MAKEDNIEMQGTVLETLPNTMFRVELENGHVVTAHISGKMRKNYIRILTGDKVTVELTPYDLSKGRIVFRSR'],\n",
        "            ['P53_HUMAN'    ,'Human',           'MEEPQSDPSVEPPLSQETFSDLWKLLPENNVLSPLPSQAMDDLMLSPDDIEQWFTEDPGPDEAPRMPEAAPRVAPAPAAPTPAAPAPAPSWPLSSSVPSQKTYQGSYGFRLGFLHSGTAKSVTCTYSPALNKMFCQLAKTCPVQLWVDSTPPPGTRVRAMAIYKQSQHMTEVVRRCPHHERCSDSDGLAPPQHLIRVEGNLRVEYLDDRNTFRHSVVVPYEPPEVGSDCTTIHYNYMCNSSCMGGMNRRPILTIITLEDSSGNLLGRNSFEVRVCACPGRDRRTEEENLRKKGEPHHELPPGSTKRALPNNTSSSPQPKKKPLDGEYFTLQIRGRERFEMFRELNEALELKDAQAGKEPGGSRAHSSHLKSKKGQSTSRHKKLMFKTEGPDSD'],\n",
        "            ['BLAT_ECOLX'\t  ,'Prokaryote',      'MSIQHFRVALIPFFAAFCLPVFAHPETLVKVKDAEDQLGARVGYIELDLNSGKILESFRPEERFPMMSTFKVLLCGAVLSRVDAGQEQLGRRIHYSQNDLVEYSPVTEKHLTDGMTVRELCSAAITMSDNTAANLLLTTIGGPKELTAFLHNMGDHVTRLDRWEPELNEAIPNDERDTTMPAAMATTLRKLLTGELLTLASRQQLIDWMEADKVAGPLLRSALPAGWFIADKSGAGERGSRGIIAALGPDGKPSRIVVIYTTGSQATMDERNRQIAEIGASLIKHW'],\n",
        "            ['BRCA1_HUMAN'\t,'Human',           'MDLSALRVEEVQNVINAMQKILECPICLELIKEPVSTKCDHIFCKFCMLKLLNQKKGPSQCPLCKNDITKRSLQESTRFSQLVEELLKIICAFQLDTGLEYANSYNFAKKENNSPEHLKDEVSIIQSMGYRNRAKRLLQSEPENPSLQETSLSVQLSNLGTVRTLRTKQRIQPQKTSVYIELGSDSSEDTVNKATYCSVGDQELLQITPQGTRDEISLDSAKKAACEFSETDVTNTEHHQPSNNDLNTTEKRAAERHPEKYQGSSVSNLHVEPCGTNTHASSLQHENSSLLLTKDRMNVEKAEFCNKSKQPGLARSQHNRWAGSKETCNDRRTPSTEKKVDLNADPLCERKEWNKQKLPCSENPRDTEDVPWITLNSSIQKVNEWFSRSDELLGSDDSHDGESESNAKVADVLDVLNEVDEYSGSSEKIDLLASDPHEALICKSERVHSKSVESNIEDKIFGKTYRKKASLPNLSHVTENLIIGAFVTEPQIIQERPLTNKLKRKRRPTSGLHPEDFIKKADLAVQKTPEMINQGTNQTEQNGQVMNITNSGHENKTKGDSIQNEKNPNPIESLEKESAFKTKAEPISSSISNMELELNIHNSKAPKKNRLRRKSSTRHIHALELVVSRNLSPPNCTELQIDSCSSSEEIKKKKYNQMPVRHSRNLQLMEGKEPATGAKKSNKPNEQTSKRHDSDTFPELKLTNAPGSFTKCSNTSELKEFVNPSLPREEKEEKLETVKVSNNAEDPKDLMLSGERVLQTERSVESSSISLVPGTDYGTQESISLLEVSTLGKAKTEPNKCVSQCAAFENPKGLIHGCSKDNRNDTEGFKYPLGHEVNHSRETSIEMEESELDAQYLQNTFKVSKRQSFAPFSNPGNAEEECATFSAHSGSLKKQSPKVTFECEQKEENQGKNESNIKPVQTVNITAGFPVVGQKDKPVDNAKCSIKGGSRFCLSSQFRGNETGLITPNKHGLLQNPYRIPPLFPIKSFVKTKCKKNLLEENFEEHSMSPEREMGNENIPSTVSTISRNNIRENVFKEASSSNINEVGSSTNEVGSSINEIGSSDENIQAELGRNRGPKLNAMLRLGVLQPEVYKQSLPGSNCKHPEIKKQEYEEVVQTVNTDFSPYLISDNLEQPMGSSHASQVCSETPDDLLDDGEIKEDTSFAENDIKESSAVFSKSVQKGELSRSPSPFTHTHLAQGYRRGAKKLESSEENLSSEDEELPCFQHLLFGKVNNIPSQSTRHSTVATECLSKNTEENLLSLKNSLNDCSNQVILAKASQEHHLSEETKCSASLFSSQCSELEDLTANTNTQDPFLIGSSKQMRHQSESQGVGLSDKELVSDDEERGTGLEENNQEEQSMDSNLGEAASGCESETSVSEDCSGLSSQSDILTTQQRDTMQHNLIKLQQEMAELEAVLEQHGSQPSNSYPSIISDSSALEDLRNPEQSTSEKAVLTSQKSSEYPISQNPEGLSADKFEVSADSSTSKNKEPGVERSSPSKCPSLDDRWYMHSCSGSLQNRNYPSQEELIKVVDVEEQQLEESGPHDLTETSYLPRQDLEGTPYLESGISLFSDDPESDPSEDRAPESARVGNIPSSTSALKVPQLKVAESAQSPAAAHTTDTAGYNAMEESVSREKPELTASTERVNKRMSMVVSGLTPEEFMLVYKFARKHHITLTNLITEETTHVVMKTDAEFVCERTLKYFLGIAGGKWVVSYFWVTQSIKERKMLNEHDFEVRGDVVNGRNHQGPKRARESQDRKIFRGLEICCYGPFTNMPTDQLEWMVQLCGASVVKELSSFTLGTGVHPIVVVQPDAWTEDNGFHAIGQMCEAPVVTREWVLDSVALYQCQELDTYLIPQIPHSHY'],\n",
        "            ['CALM1_HUMAN'\t,'Human',           'MADQLTEEQIAEFKEAFSLFDKDGDGTITTKELGTVMRSLGQNPTEAELQDMINEVDADGNGTIDFPEFLTMMARKMKDTDSEEEIREAFRVFDKDGNGYISAAELRHVMTNLGEKLTDEEVDEMIREADIDGDGQVNYEEFVQMMTAK'],\n",
        "            ['CCDB_ECOLI'\t  ,'Prokaryote',\t    'MQFKVYTYKRESRYRLFVDVQSDIIDTPGRRMVIPLASARLLSDKVSRELYPVVHIGDESWRMMTTDMASVPVSVIGEEVADLSHRENDIKNAINLMFWGI'],\n",
        "            ['GFP_AEQVI'\t  ,'Other eukaryote', 'MSKGEELFTGVVPILVELDGDVNGHKFSVSGEGEGDATYGKLTLKFICTTGKLPVPWPTLVTTLSYGVQCFSRYPDHMKQHDFFKSAMPEGYVQERTIFFKDDGNYKTRAEVKFEGDTLVNRIELKGIDFKEDGNILGHKLEYNYNSHNVYIMADKQKNGIKVNFKIRHNIEDGSVQLADHYQQNTPIGDGPVLLPDNHYLSTQSALSKDPNEKRDHMVLLEFVTAAGITHGMDELYK'],\n",
        "            ['GRB2_HUMAN'\t  ,'Human',           'MEAIAKYDFKATADDELSFKRGDILKVLNEECDQNWYKAELNGKDGFIPKNYIEMKPHPWFFGKIPRAKAEEMLSKQRHDGAFLIRESESAPGDFSLSVKFGNDVQHFKVLRDGAGKYFLWVVKFNSLNELVDYHRSTSVSRNQQIFLRDIEQVPQQPTYVQALFDFDPQEDGELGFRRGDFIHVMDNSDPNWWKGACHGQTGMFPRNYVTPVNRNV'],\n",
        "            ['HSP82_YEAST'\t,'Eukaryote\t',      'MASETFEFQAEITQLMSLIINTVYSNKEIFLRELISNASDALDKIRYKSLSDPKQLETEPDLFIRITPKPEQKVLEIRDSGIGMTKAELINNLGTIAKSGTKAFMEALSAGADVSMIGQFGVGFYSLFLVADRVQVISKSNDDEQYIWESNAGGSFTVTLDEVNERIGRGTILRLFLKDDQLEYLEEKRIKEVIKRHSEFVAYPIQLVVTKEVEKEVPIPEEEKKDEEKKDEEKKDEDDKKPKLEEVDEEEEKKPKTKKVKEEVQEIEELNKTKPLWTRNPSDITQEEYNAFYKSISNDWEDPLYVKHFSVEGQLEFRAILFIPKRAPFDLFESKKKKNNIKLYVRRVFITDEAEDLIPEWLSFVKGVVDSEDLPLNLSREMLQQNKIMKVIRKNIVKKLIEAFNEIAEDSEQFEKFYSAFSKNIKLGVHEDTQNRAALAKLLRYNSTKSVDELTSLTDYVTRMPEHQKNIYYITGESLKAVEKSPFLDALKAKNFEVLFLTDPIDEYAFTQLKEFEGKTLVDITKDFELEETDEEKAEREKEIKEYEPLTKALKEILGDQVEKVVVSYKLLDAPAAIRTGQFGWSANMERIMKAQALRDSSMSSYMSSKKTFEISPKSPIIKELKKRVDEGGAQDKTVKDLTKLLYETALLTSGFSLDEPTSFASRINRLISLGLNIDEDEETETAPEASTAAPVEEVPADTEMEEVD'],\n",
        "            ['IF1_ECOLI'\t  ,'Prokaryote',\t    'MAKEDNIEMQGTVLETLPNTMFRVELENGHVVTAHISGKMRKNYIRILTGDKVTVELTPYDLSKGRIVFRSR'],\n",
        "            ['KCNH2_HUMAN'\t,'Human',           'MPVRRGHVAPQNTFLDTIIRKFEGQSRKFIIANARVENCAVIYCNDGFCELCGYSRAEVMQRPCTCDFLHGPRTQRRAAAQIAQALLGAEERKVEIAFYRKDGSCFLCLVDVVPVKNEDGAVIMFILNFEVVMEKDMVGSPAHDTNHRGPPTSWLAPGRAKTFRLKLPALLALTARESSVRSGGAGGAGAPGAVVVDVDLTPAAPSSESLALDEVTAMDNHVAGLGPAEERRALVGPGSPPRSAPGQLPSPRAHSLNPDASGSSCSLARTRSRESCASVRRASSADDIEAMRAGVLPPPPRHASTGAMHPLRSGLLNSTSDSDLVRYRTISKIPQITLNFVDLKGDPFLASPTSDREIIAPKIKERTHNVTEKVTQVLSLGADVLPEYKLQAPRIHRWTILHYSPFKAVWDWLILLLVIYTAVFTPYSAAFLLKETEEGPPATECGYACQPLAVVDLIVDIMFIVDILINFRTTYVNANEEVVSHPGRIAVHYFKGWFLIDMVAAIPFDLLIFGSGSEELIGLLKTARLLRLVRVARKLDRYSEYGAAVLFLLMCTFALIAHWLACIWYAIGNMEQPHMDSRIGWLHNLGDQIGKPYNSSGLGGPSIKDKYVTALYFTFSSLTSVGFGNVSPNTNSEKIFSICVMLIGSLMYASIFGNVSAIIQRLYSGTARYHTQMLRVREFIRFHQIPNPLRQRLEEYFQHAWSYTNGIDMNAVLKGFPECLQADICLHLNRSLLQHCKPFRGATKGCLRALAMKFKTTHAPPGDTLVHAGDLLTALYFISRGSIEILRGDVVVAILGKNDIFGEPLNLYARPGKSNGDVRALTYCDLHKIHRDDLLEVLDMYPEFSDHFWSSLEITFNLRDTNMIPGSPGSTELEGGFSRQRKRKLSFRRRTDKDTEQPGEVSALGPGRAGAGPSSRGRPGGPWGESPSSGPSSPESSEDEGPGRSSSPLRLVPFSSPRPPGEPPGGEPLMEDCEKSSDTCNPLSGAFSGVSNIFSFWGDSRGRQYQELPRCPAPTPSLLNIPLSSPGRRPRGDVESRLDALQRQLNRLETRLSADMATVLQLLQRQMTLVPPAYSAVTTPGPGPTSTSPLLPVSPLPTLTLDSLSQVSQFMACEELPPGAPELPQEGPTRRLSLPGQLGALTSQPLHRHGSDPGS'],\n",
        "            ['KKA2_KLEPN'\t  ,'Prokaryote',\t    'MIEQDGLHAGSPAAWVERLFGYDWAQQTIGCSDAAVFRLSAQGRPVLFVKTDLSGALNELQDEAARLSWLATTGVPCAAVLDVVTEAGRDWLLLGEVPGQDLLSSHLAPAEKVSIMADAMRRLHTLDPATCPFDHQAKHRIERARTRMEAGLVDQDDLDEEHQGLAPAELFARLKARMPDGEDLVVTHGDACLPNIMVENGRFSGFIDCGRLGVADRYQDIALATRDIAEELGGEWADRFLVLYGIAAPDSQRIAFYRLLDEFF'],\n",
        "            ['MSH2_HUMAN'\t  ,'Human',           'MAVQPKETLQLESAAEVGFVRFFQGMPEKPTTTVRLFDRGDFYTAHGEDALLAAREVFKTQGVIKYMGPAGAKNLQSVVLSKMNFESFVKDLLLVRQYRVEVYKNRAGNKASKENDWYLAYKASPGNLSQFEDILFGNNDMSASIGVVGVKMSAVDGQRQVGVGYVDSIQRKLGLCEFPDNDQFSNLEALLIQIGPKECVLPGGETAGDMGKLRQIIQRGGILITERKKADFSTKDIYQDLNRLLKGKKGEQMNSAVLPEMENQVAVSSLSAVIKFLELLSDDSNFGQFELTTFDFSQYMKLDIAAVRALNLFQGSVEDTTGSQSLAALLNKCKTPQGQRLVNQWIKQPLMDKNRIEERLNLVEAFVEDAELRQTLQEDLLRRFPDLNRLAKKFQRQAANLQDCYRLYQGINQLPNVIQALEKHEGKHQKLLLAVFVTPLTDLRSDFSKFQEMIETTLDMDQVENHEFLVKPSFDPNLSELREIMNDLEKKMQSTLISAARDLGLDPGKQIKLDSSAQFGYYFRVTCKEEKVLRNNKNFSTVDIQKNGVKFTNSKLTSLNEEYTKNKTEYEEAQDAIVKEIVNISSGYVEPMQTLNDVLAQLDAVVSFAHVSNGAPVPYVRPAILEKGQGRIILKASRHACVEVQDEIAFIPNDVYFEKDKQMFHIITGPNMGGKSTYIRQTGVIVLMAQIGCFVPCESAEVSIVDCILARVGAGDSQLKGVSTFMAEMLETASILRSATKDSLIIIDELGRGTSTYDGFGLAWAISEYIATKIGAFCMFATHFHELTALANQIPTVNNLHVTALTTEETLTMLYQVKKGVCDQSFGIHVAELANFPKHVIECAKQKALELEEFQYIGESQGYDIMEPAAKKCYLEREQGEKIIQEFLSKVKQMPFTEMSEENITIKLKQLKAEVIAKNNSFVNEIISRIKVTT'],\n",
        "            ['PABP_YEAST'   ,'Other eukaryote', 'MADITDKTAEQLENLNIQDDQKQAATGSESQSVENSSASLYVGDLEPSVSEAHLYDIFSPIGSVSSIRVCRDAITKTSLGYAYVNFNDHEAGRKAIEQLNYTPIKGRLCRIMWSQRDPSLRKKGSGNIFIKNLHPDIDNKALYDTFSVFGDILSSKIATDENGKSKGFGFVHFEEEGAAKEAIDALNGMLLNGQEIYVAPHLSRKERDSQLEETKAHYTNLYVKNINSETTDEQFQELFAKFGPIVSASLEKDADGKLKGFGFVNYEKHEDAVKAVEALNDSELNGEKLYVGRAQKKNERMHVLKKQYEAYRLEKMAKYQGVNLFVKNLDDSVDDEKLEEEFAPYGTITSAKVMRTENGKSKGFGFVCFSTPEEATKAITEKNQQIVAGKPLYVAIAQRKDVRRSQLAQQIQARNQMRYQQATAAAAAAAAGMPGQFMPPMFYGVMPPRGVPFNGPNPQQMNPMGGMPKNGMPPQFRNGPVYGVPPQGGFPRNANDNNQFYQQKQRQALGEQLYKKVSAKTSNEEAAGKITGMILDLPPQEVFPLLESDELFEQHYKEASAAYESFKKEQEQQTEQA'],\n",
        "            ['P53_HUMAN'\t  ,'Human',           'MEEPQSDPSVEPPLSQETFSDLWKLLPENNVLSPLPSQAMDDLMLSPDDIEQWFTEDPGPDEAPRMPEAAPRVAPAPAAPTPAAPAPAPSWPLSSSVPSQKTYQGSYGFRLGFLHSGTAKSVTCTYSPALNKMFCQLAKTCPVQLWVDSTPPPGTRVRAMAIYKQSQHMTEVVRRCPHHERCSDSDGLAPPQHLIRVEGNLRVEYLDDRNTFRHSVVVPYEPPEVGSDCTTIHYNYMCNSSCMGGMNRRPILTIITLEDSSGNLLGRNSFEVRVCACPGRDRRTEEENLRKKGEPHHELPPGSTKRALPNNTSSSPQPKKKPLDGEYFTLQIRGRERFEMFRELNEALELKDAQAGKEPGGSRAHSSHLKSKKGQSTSRHKKLMFKTEGPDSD'],\n",
        "            ['PTEN_HUMAN'\t  ,'Human',           'MTAIIKEIVSRNKRRYQEDGFDLDLTYIYPNIIAMGFPAERLEGVYRNNIDDVVRFLDSKHKNHYKIYNLCAERHYDTAKFNCRVAQYPFEDHNPPQLELIKPFCEDLDQWLSEDDNHVAAIHCKAGKGRTGVMICAYLLHRGKFLKAQEALDFYGEVRTRDKKGVTIPSQRRYVYYYSYLLKNHLDYRPVALLFHKMMFETIPMFSGGTCNPQFVVCQLKVKIYSSNSGPTRREDKFMYFEFPQPLPVCGDIKVEFFHKQNKMLKKDKMFHFWVNTFFIPGPEETSEKVENGSLCDQEIDSICSIERADNDKEYLVLTLTKNDLDKANKDKANRYFSPNFKVKLYFTKTVEEPSNPEASSSTSVTPDVSDNEPDHYRYSDTTDSDPENEPFDEDQHTQITKV'],\n",
        "            ['RL40A_YEAST'\t,'Eukaryote\t',      'MQIFVKTLTGKTITLEVESSDTIDNVKSKIQDKEGIPPDQQRLIFAGKQLEDGRTLSDYNIQKESTLHLVLRLRGGIIEPSLKALASKYNCDKSVCRKCYARLPPRATNCRKRKCGHTNQLRPKKKLK'],\n",
        "            ['SCN5A_HUMAN'\t,'Human',           'MANFLLPRGTSSFRRFTRESLAAIEKRMAEKQARGSTTLQESREGLPEEEAPRPQLDLQASKKLPDLYGNPPQELIGEPLEDLDPFYSTQKTFIVLNKGKTIFRFSATNALYVLSPFHPIRRAAVKILVHSLFNMLIMCTILTNCVFMAQHDPPPWTKYVEYTFTAIYTFESLVKILARGFCLHAFTFLRDPWNWLDFSVIIMAYTTEFVDLGNVSALRTFRVLRALKTISVISGLKTIVGALIQSVKKLADVMVLTVFCLSVFALIGLQLFMGNLRHKCVRNFTALNGTNGSVEADGLVWESLDLYLSDPENYLLKNGTSDVLLCGNSSDAGTCPEGYRCLKAGENPDHGYTSFDSFAWAFLALFRLMTQDCWERLYQQTLRSAGKIYMIFFMLVIFLGSFYLVNLILAVVAMAYEEQNQATIAETEEKEKRFQEAMEMLKKEHEALTIRGVDTVSRSSLEMSPLAPVNSHERRSKRRKRMSSGTEECGEDRLPKSDSEDGPRAMNHLSLTRGLSRTSMKPRSSRGSIFTFRRRDLGSEADFADDENSTAGESESHHTSLLVPWPLRRTSAQGQPSPGTSAPGHALHGKKNSTVDCNGVVSLLGAGDPEATSPGSHLLRPVMLEHPPDTTTPSEEPGGPQMLTSQAPCVDGFEEPGARQRALSAVSVLTSALEELEESRHKCPPCWNRLAQRYLIWECCPLWMSIKQGVKLVVMDPFTDLTITMCIVLNTLFMALEHYNMTSEFEEMLQVGNLVFTGIFTAEMTFKIIALDPYYYFQQGWNIFDSIIVILSLMELGLSRMSNLSVLRSFRLLRVFKLAKSWPTLNTLIKIIGNSVGALGNLTLVLAIIVFIFAVVGMQLFGKNYSELRDSDSGLLPRWHMMDFFHAFLIIFRILCGEWIETMWDCMEVSGQSLCLLVFLLVMVIGNLVVLNLFLALLLSSFSADNLTAPDEDREMNNLQLALARIQRGLRFVKRTTWDFCCGLLRQRPQKPAALAAQGQLPSCIATPYSPPPPETEKVPPTRKETRFEEGEQPGQGTPGDPEPVCVPIAVAESDTDDQEEDEENSLGTEEESSKQQESQPVSGGPEAPPDSRTWSQVSATASSEAEASASQADWRQQWKAEPQAPGCGETPEDSCSEGSTADMTNTAELLEQIPDLGQDVKDPEDCFTEGCVRRCPCCAVDTTQAPGKVWWRLRKTCYHIVEHSWFETFIIFMILLSSGALAFEDIYLEERKTIKVLLEYADKMFTYVFVLEMLLKWVAYGFKKYFTNAWCWLDFLIVDVSLVSLVANTLGFAEMGPIKSLRTLRALRPLRALSRFEGMRVVVNALVGAIPSIMNVLLVCLIFWLIFSIMGVNLFAGKFGRCINQTEGDLPLNYTIVNNKSQCESLNLTGELYWTKVKVNFDNVGAGYLALLQVATFKGWMDIMYAAVDSRGYEEQPQWEYNLYMYIYFVIFIIFGSFFTLNLFIGVIIDNFNQQKKKLGGQDIFMTEEQKKYYNAMKKLGSKKPQKPIPRPLNKYQGFIFDIVTKQAFDVTIMFLICLNMVTMMVETDDQSPEKINILAKINLLFVAIFTGECIVKLAALRHYYFTNSWNIFDFVVVILSIVGTVLSDIIQKYFFSPTLFRVIRLARIGRILRLIRGAKGIRTLLFALMMSLPALFNIGLLLFLVMFIYSIFGMANFAYVKWEAGIDDMFNFQTFANSMLCLFQITTSAGWDGLLSPILNTGPPYCDPTLPNSNGSRGDCGSPAVGILFFTTYIIISFLIVVNMYIAIILENFSVATEESTEPLSEDDFDMFYEIWEKFDPEATQFIEYSVLSDFADALSEPLRIAKPNQISLINMDLPMVSGDRIHCMDILFAFTKRVLGESGEMDALKIQMEEKFMAANPSKISYEPITTTLRRKHEEVSAMVIQRAFRRHLLQRSLKHASFLFRQQAGSGLSEEDAPEREGLIAYVMSENFSRPLGPPSSSSISSTSFPPSYDSVTRATSDNLQVRGSDYSHSEDLADFPPSPDRDRESIV'],\n",
        "            ['SUMO1_HUMAN'  ,'Human',           'MSDQEAKPSTEDLGDKKEGEYIKLKVIGQDSSEIHFKVKMTTHLKKLKESYCQRQGVPMNSLRFLFEGQRIADNHTPKELGMEEEDVIEVYQEQTGGHSTV']\n",
        "        ],\n",
        "    )\n",
        "    gr.Markdown(\"<br>\")\n",
        "    gr.Markdown(\"# Fitness predictions for all single amino acid substitutions in mutation range\")\n",
        "    gr.Markdown(\"Inference may take a few seconds for short proteins & mutation ranges to several minutes for longer ones\")\n",
        "    output_image = gr.Gallery(label=\"Fitness predictions for all single amino acid substitutions in mutation range\",type=\"filepath\") #Using Gallery to break down large scoring matrices into smaller images\n",
        "    \n",
        "    output_recommendations = gr.Textbox(label=\"Mutation recommendations\")\n",
        "    \n",
        "    clear_button.click(\n",
        "        inputs = [protein_sequence_input,mutation_range_start,mutation_range_end],\n",
        "        outputs = [protein_sequence_input,mutation_range_start,mutation_range_end],\n",
        "        fn=clear_inputs\n",
        "    )\n",
        "    run_button.click(\n",
        "        fn=score_and_create_matrix_all_singles,\n",
        "        inputs=[protein_sequence_input,mutation_range_start,mutation_range_end,model_size_selection,scoring_mirror,batch_size_inference],\n",
        "        outputs=[output_image,output_recommendations],\n",
        "    )\n",
        "    gr.Markdown(\"# Mutate the starting protein sequence\")\n",
        "    with gr.Row():\n",
        "        mutation_triplet = gr.Textbox(lines=1,label=\"Selected mutation\", placeholder = \"Input the mutation triplet for the selected mutation (eg., M1A)\")\n",
        "    mutate_button = gr.Button(value=\"Apply mutation to starting protein\", variant=\"primary\")\n",
        "    mutated_protein_sequence = gr.Textbox(lines=1,label=\"Mutated protein sequence\")\n",
        "    mutate_button.click(\n",
        "        fn = get_mutated_protein,\n",
        "        inputs = [protein_sequence_input,mutation_triplet],\n",
        "        outputs = mutated_protein_sequence\n",
        "    )\n",
        "    gr.Markdown(\"<p>You may now use the output mutated sequence above as the starting sequence for another round of in silico directed evolution.</p>\")\n",
        "    gr.Markdown(\"For more information about the Tranception model, please refer to our paper below:\")\n",
        "    gr.Markdown(\"<p><b>Tranception: Protein Fitness Prediction with Autoregressive Transformers and Inference-time Retrieval</b><br>Pascal Notin, Mafalda Dias, Jonathan Frazer, Javier Marchena-Hurtado, Aidan N. Gomez, Debora S. Marks<sup>*</sup>, Yarin Gal<sup>*</sup><br><sup>* equal senior authorship</sup></p>\")\n",
        "    gr.Markdown(\"Links: <a href='https://proceedings.mlr.press/v162/notin22a.html' target='_blank'>Paper</a>  <a href='https://github.com/OATML-Markslab/Tranception' target='_blank'>Code</a>  <a href='https://sites.google.com/view/proteingym/substitutions' target='_blank'>ProteinGym</a>\")\n",
        "\n",
        "tranception_design.launch(debug=True,share=True)"
      ]
    }
  ],
  "metadata": {
    "accelerator": "GPU",
    "colab": {
      "collapsed_sections": [],
      "provenance": []
    },
    "gpuClass": "standard",
    "kernelspec": {
      "display_name": "Python 3",
      "name": "python3"
    },
    "language_info": {
      "name": "python"
    }
  },
  "nbformat": 4,
  "nbformat_minor": 0
}
